### ISSUE TITLE
Inconsistency between ConvGNT and LinearGNT

### ISSUE THREAD

**Issue by Erotemic** [shibhansh/loss-of-plasticity#2]
The linear gnt in update optim params: 
https://github.com/shibhansh/loss-of-plasticity/blob/main/lop/algos/gnt.py#L216

And corresponding conv update optim params:
https://github.com/shibhansh/loss-of-plasticity/blob/main/lop/algos/gnt.py#L216

Are inconsistent. The conv variant does not zero the exp_avg state for the weights, but it does for the bias. There are several other inconsistencies I've noticed between them.

FWIW I am working on a modified implementation that uses the same code path for both conv and linear variants. Ultimately I want to try this out on a complex transformer. My WIP variant is here: https://github.com/Erotemic/loss-of-plasticity/blob/main/lop/algos/gen_and_test.py if there is interest. One of the main places I could use input is what to do when a layer's output connects to more than one successor layers.

**Comment by shibhansh** [comment]
Hey, sorry about the late reply. For some reason, I didn't get notification about the issues on this repository.
You are correct. There is a bug where the conv variant does not zero the exp_avg state for the weights.

I'm glad to see that you are interested in extending the implementation to more complex networks. If you're still interested in it, we can chat later to see what needs to be done.

**Comment by Erotemic** [comment]
Yes, I'm interested in working on this. I've been using a restricted form of this to annecdotal success. I haven't measured if GnT is improving my results, but I do feel like my networks are much less prone to overfitting. I'll have to think about how to setup that experiment.

My current approach uses torchview (in a very hacky way) to trace out a graph of the network structure. I've also had the author of TorchLens reach out to me and indicate that torchlens might work better for this. I haven't looked into it yet. Unhacking the automatic extraction and converting it to a networkx graph is the first major improvement my implementation extension needs.

Given the correct graph structure of what weights are connected to each activation the problem becomes much more academic. I've been studying the continual learning paper, and I think the extension might not be too hard. It just needs to be phrased generally as a graph problem, but the underlying equations should stay roughly the same.

The other implementation detail that I have to work around is accumulation of gradients to simulate larger batch sizes. I had code which I thought would maintain the utility estimates over multiple gradient accumulations, but I had to disable it because it slowed down everything by over 10x. Thinking about it, I'm not 100% sure I was approaching it correctly either.

In any case, I'm looking forward to working with you on this.

**Comment by shibhansh** [comment]
Hey, hope you had good holidays. Please send me your email, mine is dohare@ualberta.ca, we should start our conversation over email. I feel there are some good ideas in the message you sent, but they'd have to be tested more thoroughly.
